src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,495 @@
+          record (toTableName tn) d key row
+
+writeRow :: (AsString k, ToJSON v) => WriteType -> Domain k v -> k -> v -> BlockHandler SQLiteEnv ()
+writeRow wt d k v = do
+    case d of
+      (UserTables _) -> writeUser wt d k v
+      _ -> writeSys wt d k v
+
+doKeys :: (IsString k, AsString k) => Domain k v -> BlockHandler SQLiteEnv [k]
+doKeys d = do
+  let tn = domainTableName d
+  version <- gets (_bvVersion . _bsBlockVersion)
+  ks <- callDb "doKeys" $ \db -> qry db
+            ("SELECT rowkey FROM [" <> tn <> "] WHERE version = ?;")
+            [SInt (fromIntegral version)]
+            [RText]
+  forM ks $ \row -> do
+    case row of
+      [SText (Utf8 k)] -> return $ fromString $ toS k
+      _ -> internalError "doKeys: The impossible happpened."
+{-# INLINE doKeys #-}
+
+doTxIds :: TableName -> TxId -> BlockHandler SQLiteEnv [TxId]
+doTxIds (TableName tn) (TxId tid) = do
+  version <- gets (_bvVersion . _bsBlockVersion)
+  rows <- callDb "doTxIds" $ \db ->
+    qry db ("SELECT txid FROM [" <> Utf8 (toS tn) <> "] WHERE version = ? AND txid > ?")
+    [SInt (fromIntegral version), SInt (fromIntegral tid)]
+    [RInt]
+  forM rows $ \row -> do
+    case row of
+      [SInt tid'] -> return $ TxId (fromIntegral tid')
+      _ -> internalError "doTxIds: the impossible happened"
+{-# INLINE doTxIds #-}
+
+record :: (AsString k, ToJSON v) => TableName -> Domain k v -> k -> v -> BlockHandler SQLiteEnv ()
+record tt d k v = bsTxRecord %= M.insertWith (flip (++)) tt [TxLog (asString d) (asString k) (toJSON v)]
+
+doCreateUserTable :: TableName -> ModuleName -> BlockHandler SQLiteEnv ()
+doCreateUserTable tn@(TableName _) mn = do
+  bs <- gets _bsBlockVersion
+  createUserTable tn bs
+  let uti = UserTableInfo mn
+  {- until I find a better way to do this -}
+  bsTxRecord %= M.insertWith (flip (++)) "SYS:usertables" [TxLog "SYS:usertables" (asString tn) (toJSON uti)]
+{-# INLINE doCreateUserTable #-}
+
+doRollback :: BlockHandler SQLiteEnv ()
+doRollback = do
+  tryAny (rollbackSavepoint PactDbTransaction) >>= \case
+    Left (SomeException err) -> logError $ "doRollback: " ++ show err
+    Right _ -> return ()
+  resetTemp
+
+doBegin :: ExecutionMode -> BlockHandler SQLiteEnv (Maybe TxId)
+doBegin m = do
+  use bsMode >>= \m' -> case m' of
+    Just {} -> do
+      logError "beginTx: In transaction, rolling back"
+      doRollback
+    Nothing -> return ()
+  resetTemp
+  beginSavepoint PactDbTransaction
+  bsMode .= Just m
+  case m of
+    Transactional -> Just <$> use bsTxId
+    Local -> pure Nothing
+{-# INLINE doBegin #-}
+
+resetTemp :: BlockHandler SQLiteEnv ()
+resetTemp = bsMode  .= Nothing >> bsTxRecord .= M.empty
+-- WE SHOULD FLUSH THE IN-BLOCK CACHE HERE
+
+doCommit :: BlockHandler SQLiteEnv [TxLog Value]
+doCommit = use bsMode >>= \mm -> case mm  of
+  Nothing -> doRollback >> internalError "doCommit: Not in transaction"
+  Just m -> do
+    txrs <- use bsTxRecord
+
+    if m == Transactional then do
+      bsTxId += 1
```

Comment:
Would prefer not to allow a thunk in here at all, although I don't think you can accumulate a thunk chain of length >= 2. (We always force txid at each row write and block history insert). Prefer `Control.Monad.State.Strict.modify'` if possible.

Reply:
Are you sure you want to go against Ed lens code here? I mean I would probably prefer `.= succ` but we use lens all over the place. Given we're IO bound here I can't really justify this low-level of a change.

Reply:
done